api/ccapi/include/model.h

```diff
@@ -40,6 +41,30 @@ enum class ModelType {
   UNKNOWN     /** Unknown */
 };
 
+/**
+ * @brief Model saving options
+ *
+ */
+enum class ModelSaveLoadFlags {
+  MODEL_SAVE_LOAD_FLAGS_INFERENCE_PARAMS =
+    ML_TRAIN_MODEL_SAVE_LOAD_FLAGS_INFERENCE_PARAMS, /**< params required for
+                                                        inferences, eg) model
+                                                        weights */
+  MODEL_SAVE_LOAD_FLAGS_TRAINING_PARAMS =
+    ML_TRAIN_MODEL_SAVE_LOAD_FLAGS_TRAINING_PARAMS, /**< params required for
+                                                       training, eg) optimizer
+                                                       weights */
+  MODEL_SAVE_LOAD_FLAGS_INFERENCE_CONFIG =
+    1 << 2, /**< inference related model_definitions, eg) graph */
+  MODEL_SAVE_LOAD_FLAGS_TRAINING_CONFIG =
+    1 << 3, /**< training related model_definitions, eg) optimizer */
+
+  MODEL_SAVE_LOAD_FLAGS_PARAMS =
+    (1 << 0) | (1 << 1), /**< INFERENCE_PARAMS | TRAINING_PARAMS */
+  MODEL_SAVE_LOAD_FLAGS_CONFIG =
+    (1 << 2) | (1 << 3) /**< INFERENCE_CONFIG | TRAINING_CONFIG */
+};
```

Comment:
We will have multiple formats supported. How such info is expressed with this flag set?

Reply:
Sorry for the late reply.

TL; DR. Current plan is we don't expose the file format directly, but leave it as a implementation detail of this API.
In this scheme, user can't and won't need to choose format.

The reason for this design choice is...

 1. There are use case of saving only some part of the graph
 some non-exhaustive and overlapping lists are:
    - saving weights to retrain later.
    - saving weights to inference later.
    - saving graph only for inference.
    - saving optimizer and hyper params to retrain later.
  2. Every format serves only their purpose, so we can't save everything in a single format. eg) tflite support only inference related weights + graph.
  3. If user designates a format as well as what to save, they will likely to call in a ill-formed combination.
  eg) Trying to saving hyperparams with *.tflite format.

 So, we decided user shouldn't know about the format directly but rely on the api, if they want to save now and restore the graph later.
If, we need to save a model in a specified, explicit format, we will have to support `ml_train_model_export`.

| Version           | call                                                                                             | Generated File                          |
|-------------------|--------------------------------------------------------------------------------------------------|-----------------------------------------|
| Tizen 6.5         | save(prefix, FLAGS_TRAIN_PARAM)                                                                  | ${prefix}.bin                           |
|                   | save(prefix, FLAGS_TRAIN_PARAM \| FLAGS_INFER_PARAM)                                             | ${prefix}.bin                           |
|                   | save(prefix, FLAGS_INFER_PARAM)                                                                  | ${prefix}.bin                           |
| Tizen near future | save(prefix, FLAGS_TRAIN_PARAM \| FLAGS_TRAIN_CONFIG)                                            | ${prefix}.bin, ${prefix}.ini             |
|                   | save(prefix, FLAGS_TRAIN_PARAM \| FLAGS_INFER_PARAM \| FLAGS_TRAIN_CONFIG \| FLAGS_INFER_CONFIG) | ${prefix}.bin, ${prefix}.ini             |
|       | save(prefix, FLAGS_INFER_PARAM \| FLAGS_INFER_CONFIG)                                            | ${prefix}.bin ${prefix}.ini  |
| Tizen future      | save(prefix, FLAGS_INFER_PARAM \| FLAGS_INFER_CONFIG)                                            | ${prefix}.tflite (something equivalent) |

Reply:
The enum for save/load is getting too complicated. I doubt if you can handle such complicated save/load protocols.
Try to minimize the possible save file formats (best = 1). And enumerate each of the format here. I understand that we can have a lot of different combinations because there are different types of data to be stored, but users would want only one format for their needs.

E.g., save (filepath, FLAG_TFLITECOMPAT_V1); save (filepath, FLAG_PYTORCHCOMPAT_V1); save (filepath, FLAG_RAWWEIGHT_V1); ...

Reply:
Yes, I think this way is more configuration, I'll come up with suggested design 👍